packages/monitor-v2/src/balance-monitor/BalanceMonitor.ts

```diff
@@ -0,0 +1,153 @@
+  name?: string; // If not provided, address is used as name.
+  tokens: MonitoredToken[];
+}
+
+// Internal interface to hold monitored item in the BalanceMonitor.
+interface MonitoredBalance {
+  accountName: string;
+  accountAddress: string;
+  provider: Provider;
+  chainId: number;
+  networkName: string;
+  tokenSymbol: string;
+  tokenAddress?: string; // If not provided, this is monitoring native token.
+  tokenDecimals: number;
+  warnThreshold: BigNumber; // Raw balance threshold to trigger a warning.
+  errorThreshold: BigNumber; // Raw balance threshold to trigger an error.
+}
+
+function isMonitoredToken(obj: unknown): obj is MonitoredToken {
+  if (typeof obj !== "object" || obj === null) return false;
+  const token = obj as MonitoredToken;
+
+  const hasValidAddress =
+    token.address === undefined || (typeof token.address === "string" && utils.isAddress(token.address));
+  const hasValidWarnThreshold = token.warnThreshold === undefined || typeof token.warnThreshold === "number";
+  const hasValidErrorThreshold = token.errorThreshold === undefined || typeof token.errorThreshold === "number";
+
+  return hasValidAddress && hasValidWarnThreshold && hasValidErrorThreshold;
+}
+
+function isMonitoredAccount(obj: unknown): obj is MonitoredAccount {
+  if (typeof obj !== "object" || obj === null) return false;
+  const account = obj as MonitoredAccount;
+
+  const hasValidChainId = typeof account.chainId === "number";
+  const hasValidAddress = typeof account.address === "string" && utils.isAddress(account.address);
+  const hasValidName = account.name === undefined || typeof account.name === "string";
+  const hasValidTokens = Array.isArray(account.tokens) && account.tokens.every(isMonitoredToken);
+
+  return hasValidChainId && hasValidAddress && hasValidName && hasValidTokens;
+}
+
+export function isMonitoredAccountsArray(obj: unknown): obj is MonitoredAccount[] {
+  if (!Array.isArray(obj)) return false;
+  return obj.every(isMonitoredAccount);
+}
+
+export class BalanceMonitor {
+  private constructor(private readonly monitoredBalances: MonitoredBalance[]) {}
+
+  // Internal method to get raw threshold from human readable amount.
+  private static parseThreshold(threshold: number | undefined, decimals: number): BigNumber {
+    if (threshold === undefined) return BigNumber.from(0);
+    return utils.parseUnits(threshold.toString(), decimals);
+  }
+
+  static async create(
+    providers: Map<number, Provider>,
+    monitoredAccounts: MonitoredAccount[]
+  ): Promise<BalanceMonitor> {
+    const monitoredBalances: MonitoredBalance[] = [];
+
+    const monitoredBalancesPromises: Promise<void>[] = []; // Will parallelize all async calls.
+
+    for (const account of monitoredAccounts) {
+      const provider = providers.get(account.chainId);
+      if (!provider) throw new Error(`No provider for chainId ${account.chainId}`);
+
+      for (const token of account.tokens) {
+        const monitoredBalancePromise = (async () => {
```

Comment:
Minor structure question: why not set this loop up using maps like:
```ts
const monitoredBalances = await Promise.all(
  monitoredAccounts.map(account => {
    ...
    return account.tokens.map(async (token) => {
      ...
      return {
        accountName: account.name || account.address,
        ...
      });
  }).flat() // Flattens since there will be nested arrays of promises.
);
```

The advantages are:
1. Avoid creating and calling the async function in the innermost loop, which isn't super idiomatic (but totally valid).
2. Avoid keeping two arrays, one for results and the other for promises.

Reply:
Agree, this is better, fixed